File: bench/js/tree12.js
function bottom_up_tree(item, depth) {
    if (depth > 0) {
        let i = item + item;
        let left = bottom_up_tree(i - 1, depth - 1);
        let right = bottom_up_tree(i, depth - 1);
        return [item, left, right];
    } else {
        return [item];
    }
}

function item_check(tree) {
    if (tree.length == 3) {
        return tree[0] + item_check(tree[1]) - item_check(tree[2]);
    } else {
        return tree[0];
    }
}

let min_depth = 4;
let max_depth = 12;

let stretch_depth = max_depth + 1;
let stretch_tree = bottom_up_tree(0, stretch_depth);
console.log("stretch tree of depth", stretch_depth, "check:", item_check(stretch_tree));

let long_lived_tree = bottom_up_tree(0, max_depth);
let depth = min_depth;
while (depth <= max_depth) {
    let iterations = Math.pow(2, max_depth - depth + min_depth);
    let check = 0;
    let i = 1;
    while (i <= iterations) {
        check = check + item_check(bottom_up_tree(1, depth)) +
            item_check(bottom_up_tree(0 - 1, depth));
        i = i + 1;
    }
    depth = depth + 2;
    console.log(iterations * 2, "trees of depth", depth, "check:", check);
}
console.log("long lived tree of depth", max_depth, "check:", item_check(long_lived_tree));

File: bench/c/tree12.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

struct tree_s;
struct leaf_s;
typedef struct tree_s tree_s;
typedef struct leaf_s leaf_s;
typedef tree_s *tree_t;

struct leaf_s
{
    bool is_leaf;
    int num;
};

struct tree_s
{
    bool is_leaf;
    int num;
    tree_t lhs;
    tree_t rhs;
};

tree_t new_tree(tree_s value)
{
    tree_t ret = malloc(sizeof(tree_s));
    *ret = value;
    return ret;
}

tree_t new_leaf(int num)
{
    tree_t ret = malloc(sizeof(leaf_s));
    ret->is_leaf = true;
    ret->num = num;
    return ret;
}

tree_t bottom_up_tree(int item, int depth)
{
    if (depth > 0)
    {
        int i = item + item;
        depth -= 1;
        tree_t left = bottom_up_tree(i - 1, depth);
        tree_t right = bottom_up_tree(i, depth);
        return new_tree((tree_s){
            .is_leaf = false,
            .num = item,
            .lhs = left,
            .rhs = right,
        });
    }
    else
    {
        return new_leaf(item);
    }
}

int item_check(tree_t tree)
{
    if (tree->is_leaf)
    {
        return tree->num;
    }
    else
    {
        return tree->num + item_check(tree->lhs) - item_check(tree->rhs);
    }
}

void free_tree(tree_t tree)
{
    if (tree->is_leaf)
    {
        free(tree);
    }
    else
    {
        free_tree(tree->lhs);
        free_tree(tree->rhs);
        free(tree);
    }
}

int main()
{
    int n = 12;
    int mindepth = 4;
    int maxdepth = mindepth + 2;
    if (maxdepth < n)
    {
        maxdepth = n;
    }

    int stretch_depth = maxdepth + 1;
    tree_t stretch_tree = bottom_up_tree(0, stretch_depth);
    printf("%i\n", item_check(stretch_tree));
    free_tree(stretch_tree);

    tree_t long_lived_tree = bottom_up_tree(0, maxdepth);

    for (int depth = mindepth; depth < maxdepth + 1; depth += 2)
    {
        int iterations = 1 << (maxdepth - depth + mindepth);
        int check = 0;
        for (int i = 0; i < iterations; i++)
        {
            tree_t tree1 = bottom_up_tree(1, depth);
            check += item_check(tree1);
            free_tree(tree1);
            tree_t tree2 = bottom_up_tree(-1, depth);
            check += item_check(tree2);
            free_tree(tree2);
        }
        printf("%i\n", check);
    }

    printf("%i\n", item_check(long_lived_tree));
    free_tree(long_lived_tree);
    return 0;
}